report/5.Solution/DockerBuild.tex

```diff
@@ -0,0 +1,107 @@
+    FROM ubuntu:14.04
+
+    # Install required packages listed by nxtOSEK
+    RUN apt-get update
+    RUN apt-get -y install tk-dev ncurses-dev libmpfr-dev wget gzip tar software-properties-common xvfb
+\end{lstlisting} 
+
+The next part is the most obscure part, as wine is installed to be able to run Windows programs.
+This is done as the tool to build OIL files, only exists for Windows.
+NeXT Tools is an utility program that includes various utilities for use with nxtOSEK, including an enhanced OIL file compile\cite{nxttool}.
+It also includes an additional nxtOSEK requirement, texinfo, which is the official GNU documentation format\cite{texinfo}.
+
+\begin{lstlisting}[language=docker,label={lst:dockerimplementation2},caption={Installation of Wine and texinfo}]
+
+    # Install wine
+    RUN dpkg --add-architecture i386
+    RUN wget -nc https://dl.winehq.org/wine-builds/Release.key
+    RUN apt-key add Release.key
+    RUN apt-add-repository https://dl.winehq.org/wine-builds/ubuntu/
+    RUN apt-get -y install apt-transport-https
+    RUN apt-get update
+    RUN apt-get -y install --install-recommends winehq-staging
+    
+    # Download a specific version of texinfo, the one in 14.04 
+    # sources is not compatible with the expected version
+    RUN wget http://ftp.gnu.org/gnu/texinfo/texinfo-4.13.tar.gz
+    RUN gzip -dc < texinfo-4.13.tar.gz tar -xf -
+    RUN cd texinfo-4.13 \&\& ./configure \&\& make \&\& make install
+\end{lstlisting} 
+Next, the ARM tool chain is installed.
+The tool chain is used by nxtOSEK to build for the ARM7 CPU in the NXT.
+\begin{lstlisting}[language=docker,label={lst:dockerimplementation3},caption={Building the ARM toolchain}]
+    # Build arm toolchain from nxtOSEK
+    COPY build_arm_toolchain.sh home/
+    RUN chmod 755 home/build_arm_toolchain.sh
+    RUN home/build_arm_toolchain.sh
+    
+    # Add build.sh
+    WORKDIR /home/nxtosek
+    COPY build.sh ./
+    RUN chmod 755 ./build.sh
+\end{lstlisting}
+
+Finally, the nxtOSEK files are moved to the home folder, as cleanup, before the entry point for the container is set to the location in the home folder.
+This means that whenever the Docker image is run, any arguments are passed along to the build file in the home directory.
+
+\begin{lstlisting}[language=docker,label={lst:dockerimplementation4},caption={Moving the nxtOSEK files and setting the entrypoint}]   
+    # Move nxtOSEK files to home folder
+    RUN useradd nxtosek
+    RUN usermod --password $(echo nxtosek openssl passwd -1 -stdin) nxtosek
+    RUN chown -R nxtosek ./
+    USER nxtosek
+    ENV HOME /home/nxtosek
+    ADD nxtOSEK.tar.xz ./
+    ADD ecrobot.mak nxtOSEK/ecrobot
+    
+    # Set build.sh as entrypoint to execute when executing docker container
+    ENTRYPOINT ["/home/nxtosek/build.sh"]
+    CMD []    
+\end{lstlisting}
+It is important to mention that all of the downloaded files are cached, so the setup is not run every single time, and a cached version is used instead.
```

Comment:
Eee, I don't think it is really clear from this line, what actually happens.

Each line in the dockerfile will apply layers to the docker image, these layers are cached, which means that once you build the docker image, these won't be re-executed (unless the line or any previous lines have been modified).

This is all built into a docker image, that is then executed. The image is not exactly cached, but just a docker image.

So there are 2 parts - build docker image and run docker image
The build docker image has not been done in a couple of months, meaning this file has not even been read since last build.
The execution of a docker image does not use cache.

Hope this kind of makes sense :joy:

Reply:
Den her er på linje 20 i stedet for